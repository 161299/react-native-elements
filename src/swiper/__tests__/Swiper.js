import React from 'react';
import { Animated, View } from 'react-native';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { create } from 'react-test-renderer';

import theme from '../../config/theme';
import { ThemeProvider } from '../../config';

import ThemedSwiper, { Swiper } from '../Swiper';

describe('Swiper Component', () => {
  it('should render without issues', () => {
    const component = shallow(<Swiper theme={theme} />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render fixed without issues', () => {
    const component = shallow(<Swiper theme={theme} positionFixed />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('componentWillUnmount should be called on unmount', () => {
    const component = shallow(<Swiper theme={theme} />);
    const componentWillUnmount = jest.spyOn(
      component.instance(),
      'componentWillUnmount'
    );

    component.unmount();

    expect(componentWillUnmount).toHaveBeenCalled();
  });

  it('should resize without issues', () => {
    const component = shallow(<Swiper theme={theme} />);

    component
      .instance()
      ._onLayout({ nativeEvent: { layout: { width: 50, height: 50 } } });

    expect(component.instance().state.width).toBe(50);
  });

  describe('ref behavior', () => {
    it('should correct restart autoplay', () => {
      const component = shallow(<Swiper theme={theme} autoplayTimeout={3} />);
      const startAutoplay = jest.spyOn(component.instance(), 'startAutoplay');
      const stopAutoplay = jest.spyOn(component.instance(), 'stopAutoplay');

      component.instance().stopAutoplay();
      component.instance().startAutoplay();

      expect(stopAutoplay).toHaveBeenCalled();
      expect(startAutoplay).toHaveBeenCalled();
    });

    it('should autoplay timeout method complete without issues', () => {
      const component = shallow(
        <Swiper theme={theme}>
          <View />
          <View />
        </Swiper>
      );

      component.instance()._autoplayTimeout();

      expect(component.instance().getActiveIndex()).toBe(1);
    });

    it('should go to next slide', () => {
      const component = shallow(
        <Swiper theme={theme}>
          <View />
          <View />
        </Swiper>
      );
      const instance = component.instance();
      instance.goToNext();

      expect(instance.getActiveIndex()).toBe(1);
    });

    it('should call events handlers', () => {
      const onAnimationStart = jest.fn();
      const onAnimationEnd = jest.fn();
      const onIndexChanged = jest.fn();

      const component = shallow(
        <Swiper
          theme={theme}
          initialSlide={1}
          onAnimationStart={onAnimationStart}
          onAnimationEnd={onAnimationEnd}
          onIndexChanged={onIndexChanged}
        >
          <View />
          <View />
        </Swiper>
      );
      component.instance().goTo(); // 0

      expect(onAnimationStart).toHaveBeenCalled();
      expect(onAnimationEnd).toHaveBeenCalled();
      expect(onIndexChanged).toHaveBeenCalled();
    });

    it('should go to prev slide on vertical swiper', () => {
      const component = shallow(
        <Swiper theme={theme} initialSlide={1} direction="vertical">
          <View />
          <View />
        </Swiper>
      );
      const instance = component.instance();
      instance.goToPrev();

      expect(instance.getActiveIndex()).toBe(0);
    });

    it('should go to slide by index', () => {
      const component = shallow(
        <Swiper theme={theme}>
          <View />
          <View />
        </Swiper>
      );
      const instance = component.instance();
      instance.goTo(1);

      expect(instance.getActiveIndex()).toBe(1);
    });

    it('should not go to prev', () => {
      const component = shallow(
        <Swiper theme={theme}>
          <View />
          <View />
        </Swiper>
      );
      const instance = component.instance();
      instance.goToPrev();

      expect(instance.getActiveIndex()).toBe(0);
    });

    it('should not go to next', () => {
      const component = shallow(
        <Swiper theme={theme} initialSlide={1}>
          <View />
          <View />
        </Swiper>
      );
      const instance = component.instance();
      instance.goToNext();

      expect(instance.getActiveIndex()).toBe(1);
    });
  });

  describe('PanResponder Callbacks', () => {
    it('should allow to move', () => {
      const onAnimationStart = jest.fn();

      const component = shallow(
        <Swiper theme={theme} onAnimationStart={onAnimationStart} />
      );

      expect(
        component
          .instance()
          ._getPanResponderCallbacks()
          .onMoveShouldSetPanResponderCapture(null, { dx: 6, dy: 6 })
      ).toBeTruthy();
      expect(
        component
          .instance()
          ._getPanResponderCallbacks()
          .onMoveShouldSetPanResponderCapture(null, { dx: 4, dy: 6 })
      ).toBeFalsy();

      expect(onAnimationStart).toHaveBeenCalled();
    });

    it('should deny to move by min distance', () => {
      const component = shallow(
        <Swiper
          direction="vertical"
          theme={theme}
          containerStyle={{ width: 50, height: 50 }}
        />
      );

      expect(
        component
          .instance()
          ._getPanResponderCallbacks()
          .onMoveShouldSetPanResponderCapture(null, { dx: 6, dy: 6 })
      ).toBeTruthy();
      expect(
        component
          .instance()
          ._getPanResponderCallbacks()
          .onMoveShouldSetPanResponderCapture(null, { dx: 4, dy: 4 })
      ).toBeFalsy();
    });

    it('should deny to move by disabling gestures', () => {
      const component = shallow(
        <Swiper theme={theme} gesturesEnabled={false} />
      );

      expect(
        component
          .instance()
          ._getPanResponderCallbacks()
          .onMoveShouldSetPanResponderCapture()
      ).toBeFalsy();
    });

    it('touch release handler should be called without issues', () => {
      const component = shallow(
        <Swiper theme={theme} containerStyle={{ width: 50, height: 10 }} />
      );

      expect(
        component
          .instance()
          ._getPanResponderCallbacks()
          .onPanResponderRelease(null, { x0: 0, y0: 0, moveX: 5, moveY: 0 })
      ).toBeUndefined();
    });

    it('touch release handler should skip index changing', () => {
      const component = shallow(<Swiper theme={theme} />);

      component
        .instance()
        ._onLayout({ nativeEvent: { layout: { width: 50, height: 50 } } });

      expect(
        component
          .instance()
          ._getPanResponderCallbacks()
          .onPanResponderRelease(null, { x0: 0, y0: 0, moveX: 5, moveY: 0 })
      ).toBeUndefined();
    });

    it('touch release handler on vertical swiper should be called without issues', () => {
      const component = shallow(
        <Swiper
          theme={theme}
          direction="vertical"
          containerStyle={{ width: 50, height: 50 }}
        />
      );

      expect(
        component
          .instance()
          ._getPanResponderCallbacks()
          .onPanResponderRelease(null, { x0: 0, y0: 0, moveX: 5, moveY: 6 })
      ).toBeUndefined();
    });

    it('should works without issue', () => {
      const component = shallow(<Swiper theme={theme} />);

      expect(
        component
          .instance()
          ._getPanResponderCallbacks()
          .onPanResponderTerminationRequest()
      ).toBeFalsy();
      expect(
        component
          .instance()
          ._getPanResponderCallbacks()
          .onMoveShouldSetResponderCapture()
      ).toBeTruthy();
      expect(
        component
          .instance()
          ._getPanResponderCallbacks()
          .onPanResponderGrant()
      ).toBeUndefined();
    });
  });

  it('should apply values from theme', () => {
    const testTheme = {
      Swiper: {
        direction: 'vertical',
      },
    };

    const component = create(
      <ThemeProvider theme={testTheme}>
        <ThemedSwiper />
      </ThemeProvider>
    );

    expect(
      component.root.findAllByType(Animated.View)[0].props.style.flexDirection
    ).toBe('column');
    expect(component.toJSON()).toMatchSnapshot();
  });
});
