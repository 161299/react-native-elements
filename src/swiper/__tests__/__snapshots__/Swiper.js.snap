// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swiper Component should apply values from theme 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "height": 0,
        "left": 0,
        "overflow": "hidden",
        "position": "relative",
        "top": 0,
        "width": 0,
      }
    }
  >
    <View
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flexDirection": "column",
          "height": 0,
          "left": 0,
          "position": "relative",
          "top": 0,
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
          ],
          "width": 0,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "flexDirection": "column",
          "justifyContent": "space-between",
          "padding": 10,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "borderRadius": 3,
              "opacity": 0.3,
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderColor": "#2089dc",
                  "borderRadius": 3,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 8,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": Object {
                      "color": Array [
                        208,
                        8,
                        63,
                      ],
                      "model": "hsl",
                      "valpha": 1,
                    },
                    "fontSize": 18,
                    "paddingBottom": 1,
                    "paddingTop": 2,
                    "textAlign": "center",
                  }
                }
              >
                Prev
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
            }
          }
        />
      </View>
      <View>
        <View
          style={
            Object {
              "borderRadius": 3,
              "opacity": 0.3,
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderColor": "#2089dc",
                  "borderRadius": 3,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 8,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": Object {
                      "color": Array [
                        208,
                        8,
                        63,
                      ],
                      "model": "hsl",
                      "valpha": 1,
                    },
                    "fontSize": 18,
                    "paddingBottom": 1,
                    "paddingTop": 2,
                    "textAlign": "center",
                  }
                }
              >
                Next
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Swiper Component should display autoplay swiper 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "height": 0,
        "left": 0,
        "overflow": "hidden",
        "position": "relative",
        "top": 0,
        "width": 0,
      }
    }
  >
    <AnimatedComponent
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flexDirection": "row",
          "height": 0,
          "left": 0,
          "position": "relative",
          "top": 0,
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
          ],
          "width": 0,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "left": 0,
          "padding": 10,
          "position": "absolute",
          "right": 0,
        }
      }
    >
      <View>
        <bound _renderButton
          isFirst={true}
          type="prev"
        />
      </View>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        />
      </View>
      <View>
        <bound _renderButton
          isLast={true}
          type="next"
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`Swiper Component should display loop swiper 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "height": 0,
        "left": 0,
        "overflow": "hidden",
        "position": "relative",
        "top": 0,
        "width": 0,
      }
    }
  >
    <AnimatedComponent
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flexDirection": "row",
          "height": 0,
          "left": 0,
          "position": "relative",
          "top": 0,
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
          ],
          "width": 0,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "left": 0,
          "padding": 10,
          "position": "absolute",
          "right": 0,
        }
      }
    >
      <View>
        <bound _renderButton
          isFirst={false}
          onPress={[Function]}
          type="prev"
        />
      </View>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        />
      </View>
      <View>
        <bound _renderButton
          isLast={false}
          onPress={[Function]}
          type="next"
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`Swiper Component should display vertical swiper 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "height": 0,
        "left": 0,
        "overflow": "hidden",
        "position": "relative",
        "top": 0,
        "width": 0,
      }
    }
  >
    <AnimatedComponent
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flexDirection": "column",
          "height": 0,
          "left": 0,
          "position": "relative",
          "top": 0,
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
          ],
          "width": 0,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "flexDirection": "column",
          "justifyContent": "space-between",
          "padding": 10,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View>
        <bound _renderButton
          isFirst={true}
          type="prev"
        />
      </View>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
            }
          }
        />
      </View>
      <View>
        <bound _renderButton
          isLast={true}
          type="next"
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`Swiper Component should display with black active dot 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "height": 0,
        "left": 0,
        "overflow": "hidden",
        "position": "relative",
        "top": 0,
        "width": 0,
      }
    }
  >
    <AnimatedComponent
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flexDirection": "row",
          "height": 0,
          "left": 0,
          "position": "relative",
          "top": 0,
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
          ],
          "width": 0,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "left": 0,
          "padding": 10,
          "position": "absolute",
          "right": 0,
        }
      }
    >
      <View>
        <bound _renderButton
          isFirst={true}
          type="prev"
        />
      </View>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        />
      </View>
      <View>
        <bound _renderButton
          isLast={true}
          type="next"
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`Swiper Component should render without issues 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "height": 0,
        "left": 0,
        "overflow": "hidden",
        "position": "relative",
        "top": 0,
        "width": 0,
      }
    }
  >
    <AnimatedComponent
      onMoveShouldSetResponder={[Function]}
      onMoveShouldSetResponderCapture={[Function]}
      onResponderEnd={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderReject={[Function]}
      onResponderRelease={[Function]}
      onResponderStart={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      onStartShouldSetResponderCapture={[Function]}
      style={
        Object {
          "flexDirection": "row",
          "height": 0,
          "left": 0,
          "position": "relative",
          "top": 0,
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
          ],
          "width": 0,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "left": 0,
          "padding": 10,
          "position": "absolute",
          "right": 0,
        }
      }
    >
      <View>
        <bound _renderButton
          isFirst={true}
          type="prev"
        />
      </View>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        />
      </View>
      <View>
        <bound _renderButton
          isLast={true}
          type="next"
        />
      </View>
    </View>
  </View>
</View>
`;
